### ISSUE TITLE
Jobs occasionally get stuck and can't complete even when all steps in playbook were actually done

### ISSUE THREAD

**Issue by seaman25** [ansible/awx#14033]
### Please confirm the following

- [X] I agree to follow this project's [code of conduct](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html).
- [X] I have checked the [current issues](https://github.com/ansible/awx/issues) for duplicates.
- [X] I understand that AWX is open source software provided for free and that I might not receive a timely response.
- [X] I am **NOT** reporting a (potential) security vulnerability. (These should be emailed to `security@ansible.com` instead.)

### Bug Summary

This seems to be happening to multiple job templates in our AWX instance. I haven't figured out a pattern yet but the jobs occasionally get stuck in `Running` status and can't complete as expected, and it could keep running even when all steps in the playbook were actually already done, and I have to cancel such jobs manually. 
I was able to confirm it was actually stuck (not an UI status update issue) as the ephemeral automation job pod for the target job was still running there in k8s. Sometimes the jobs might be terminated by AWX itself in 6 hours, sometimes in 8 hours or even longer. For the jobs with a shorter timeout settings in the template (e.g. 1 hour based on the business logic it has) it will just timeout.

Most of our jobs are invoking python scripts and I was able to tell that the actual steps had been done according to the logs in our logging system produced by the python scripts.

And the issue could also happen to jobs which don't invoke any python or shell script but only execute some k8s modules in playbook to create resources in k8s cluster, i.e. only some simple tasks in the playbook. And in one case the job got stuck when the job output had already printed the PLAY RECAP summary info like following:
```
PLAY RECAP *********************************************************************
localhost                  : ok=16   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### AWX version

21.9.0

### Select the relevant components

- [ ] UI
- [ ] UI (tech preview)
- [ ] API
- [ ] Docs
- [ ] Collection
- [ ] CLI
- [X] Other

### Installation method

kubernetes

### Modifications

We don't have modifications to AWX image itself, but we do use our customized Execution Environment image:
* Base image: quay.io/ansible/ansible-runner:stable-2.12-devel
* Python version: 3.8.13

### Ansible version

2.12.5

### Steps to reproduce

Just run jobs as usual and the issue could happen occasionally, recently it almost happens everyday, maybe roughly 1% to 3% chance of happening. It happens more often to long running jobs (the ones run for longer than 10-15 minutes or so).

### Expected results

Jobs can complete (either succeed or fail) when all steps in playbook are done.

### Actual results

Some jobs get stuck in Running status and the ephemeral automation job pod will keep running in k8s until it gets timeout or terminated by AWX in 6-12 hours or so, or cancelled manually.

### Additional information

I couldn't find any helpful log in any of the AWX web/task/ee containers or the automation job pod itself. Below is an example of the logs regarding a problematic job from `mro-awx-task` container in `mro-awx` pod:
```
2023-05-21 22:45:59,902 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 created
2023-05-21 22:46:00,165 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 acknowledged
2023-05-21 22:46:00,599 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 blocked by projectupdate-283033
2023-05-21 22:46:02,647 INFO     [5e7d0e67d63e49bd8db81aecef8da319] awx.analytics.job_lifecycle job-283032 blocked by projectupdate-283033
2023-05-21 22:46:06,295 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 waiting
2023-05-21 22:46:06,657 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 pre run
2023-05-21 22:46:06,917 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 preparing playbook
2023-05-21 22:46:07,113 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 running playbook
2023-05-21 22:46:11,662 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 work unit id received
2023-05-21 22:46:11,752 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 work unit id assigned
2023-05-22 00:01:20,834 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.main.commands.run_callback_receiver Starting EOF event processing for Job 283032
2023-05-22 00:01:20,868 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 post run
2023-05-22 00:01:26,570 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 finalize run
2023-05-22 00:01:26,618 WARNING  [34e9f318d1374a0fbd4becf3dfdee5e4] awx.main.dispatch job 283032 (error) encountered an error (rc=None), please see task stdout for details.
2023-05-22 00:01:27,000 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.analytics.job_lifecycle job-283032 notifications sent
```

The log entry `2023-05-22 00:01:20,834 INFO     [34e9f318d1374a0fbd4becf3dfdee5e4] awx.main.commands.run_callback_receiver Starting EOF event processing for Job 283032` happened when I cancelled that job manually.

Below is another example of the similar logs for a job which succeeded as expected, just for a comparison with above logs:
```
2023-05-21 23:00:05,278 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 created
2023-05-21 23:00:05,470 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 acknowledged
2023-05-21 23:00:05,913 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 blocked by projectupdate-283040
2023-05-21 23:00:06,280 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 blocked by projectupdate-283040
2023-05-21 23:00:06,873 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 blocked by projectupdate-283040
2023-05-21 23:00:10,951 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.main.commands.run_callback_receiver Starting EOF event processing for Job 283040
2023-05-21 23:00:10,954 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle projectupdate-283040 post run
2023-05-21 23:00:11,324 INFO     [-] awx.analytics.job_lifecycle projectupdate-283040 stats wrapup finished
2023-05-21 23:00:11,457 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle projectupdate-283040 finalize run
2023-05-21 23:00:11,791 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 waiting
2023-05-21 23:00:12,183 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 pre run
2023-05-21 23:00:12,640 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 preparing playbook
2023-05-21 23:00:12,840 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 running playbook
2023-05-21 23:00:17,824 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 work unit id received
2023-05-21 23:00:17,904 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 work unit id assigned
2023-05-21 23:20:41,863 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.main.commands.run_callback_receiver Starting EOF event processing for Job 283039
2023-05-21 23:20:41,868 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 post run
2023-05-21 23:20:42,074 INFO     [3b04b25f62c842c7bc80cc6b495a2757] awx.analytics.job_lifecycle job-283039 finalize run
2023-05-21 23:20:42,156 INFO     [-] awx.analytics.job_lifecycle job-283039 stats wrapup finished
```

**Comment by seaman25** [comment]
I'm not sure if it's a bug in AWX or something to do with the scale of data in our AWX database or anything else. Any ideas or info about similar issues will be appreciated.

**Comment by seaman25** [comment]
Btw, this issue started to appear after we upgraded our AWX from version `19.5.1` to `21.9.0` earlier this year. It never happened with `19.5.1` but that older version had another problem where jobs would be terminated by AWX itself if there's no job output for longer than 15 minutes or so, which was fixed in `21.9.0`. Not sure if the current issue I'm facing in `21.9.0` is related to the changes for fixing that problem in older versions.

**Comment by fosterseth** [comment]
which k8s version are you using?

are the automation job pods staying in a running state?

does the result_traceback in the /api/v2/jobs/<jobid> have any useful info?

also have you experienced any failures for jobs that don't run on the custom ee? wondering if it is worth building your EE based on awx-ee:latest, instead of the ansible-runner stable.

lastly, it may be worth doing an upgrade to latest and seeing if the problem goes away.

**Comment by seaman25** [comment]
Hi @fosterseth , thanks for your reply! Here are my answers to your questions:
* `which k8s version are you using?` - We have 2 AWX instances (both on 21.9.0 version), one is running on k8s 1.24 and the other is 1.25, and I've seen this issue in both AWX instances;
* `are the automation job pods staying in a running state?` - Yes, the job pods are staying in a running status without any error or further output.
* `does the result_traceback in the /api/v2/jobs/ have any useful info?`
  * For the jobs terminated by me manually (job status `Canceled`), `result_traceback` was a blank string.
  * For one of the job terminated by AWX automatically (ran for 1 hour 14 minutes, job status `Error`), `result_traceback` has much more content but all looks like the normal job output including the stdout and ansible facts, etc.
  * For another job terminated by AWX automatically (ran for 20 hours, job status `Error Job terminated due to error`, `result_traceback` has some error info as following:
```
"result_traceback": "Traceback (most recent call last):\n  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/receptor.py\", line 427, in _run_internal\n    resultsock = receptor_ctl.get_work_results(self.unit_id, return_sockfile=True)\n  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/receptorctl/socket_interface.py\", line 255, in get_work_results\n    raise RuntimeError(errmsg)\nRuntimeError: Failed to get results: unknown work unit QkLBwmYH\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/jobs.py\", line 589, in run\n    res = receptor_job.run()\n  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/receptor.py\", line 319, in run\n    res = self._run_internal(receptor_ctl)\n  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/receptor.py\", line 437, in _run_internal\n    raise RuntimeError(detail)\nRuntimeError\n",
```
* `also have you experienced any failures for jobs that don't run on the custom ee? wondering if it is worth building your EE based on awx-ee:latest, instead of the ansible-runner stable.` - All our jobs are running on the custom EE, so for a comparison last week I created a job template which uses the `awx-ee:latest` image and created a schedule to run that job every hour, I haven't experienced the failure with that job yet, I'll monitor it for a bit longer to see if it happens.
* `lastly, it may be worth doing an upgrade to latest and seeing if the problem goes away.` - Yeah, if we can't figure anything out I might upgrade the AWX version.

**Comment by happytreees** [comment]
I am also seeing this issue on AWX recently. I did not see it on my previous v1.22 cluster, however, a new v1.26 has this issue and I have been unable to get around other than ensuring the playbook runs as quickly as possible which is not ideal. 

As Seaman is seeing even when the playbook finishes it just sits there sometimes in a running state until the idle timeout I have set kills it. I had to implement the idle timeout because it would just continue running without dying indefinitely when it was not there. 

I am running:
AWX: 22.6.0
AWX-Operator: 2.5.0
Kubernetes: 1.26.6

I've also made a custom EE with the following `ansible.cfg` parameters and it still does not help any:
```
[defaults]
transport = paramiko
use_persistent_connections = true
host_key_checking = false
```

This has been especially painful for me when running template tasks in the playbook as there is no way to use async or to speed it up. 

I ran a strace on the ansible process from the worker node that it was scheduled on and while it's hanging it just repeats:
```
wait4(433, 0x7ffcceecb284, WNOHANG, NULL) = 0
select(0, NULL, NULL, NULL, {tv_sec=0, tv_usec=1000}) = 0 (Timeout)
wait4(433, 0x7ffcceecb284, WNOHANG, NULL) = 0
select(0, NULL, NULL, NULL, {tv_sec=0, tv_usec=1000}) = 0 (Timeout)
wait4(433, 0x7ffcceecb284, WNOHANG, NULL) = 0
select(0, NULL, NULL, NULL, {tv_sec=0, tv_usec=1000}) = 0 (Timeout)
wait4(433, 0x7ffcceecb284, WNOHANG, NULL) = 0
select(0, NULL, NULL, NULL, {tv_sec=0, tv_usec=1000}) = 0 (Timeout)
wait4(433, 0x7ffcceecb284, WNOHANG, NULL) = 0
select(0, NULL, NULL, NULL, {tv_sec=0, tv_usec=1000}) = 0 (Timeout)
wait4(433, 0x7ffcceecb284, WNOHANG, NULL) = 0
```

There is absolutely no movement on the playbook once it happens.

**Comment by seaman25** [comment]
Hi @happytreees Thanks for sharing your case! Below is what I've found which might have fixed / mitigated the issue, for your reference:

The issue hasn't been happening much (almost gone completely) since we've switched to a new custom EE image about 8 weeks ago. So I guess it might be related to the compatibility between AWX itself (21.9.0 in our case) and the custom EE image:

* **Base image of our old custom EE image**:
  * OS: `quay.io/ansible/ansible-runner:stable-2.12-devel` image based on `CentOS Stream 8`
  * `receptor` dependency: `quay.io/project-receptor/receptor:1.0.0a2`
Btw, when selecting the base image for this old EE image, we were referring to the official AWX EE image `0.5.0` which is quite old now.

* **Base image of our new custom EE image**:
  * OS: `quay.io/ansible/awx-ee:22.3.0` image based on `CentOS Stream 9`
  * `receptor` dependency: `quay.io/ansible/receptor:devel`

Our old custom EE image worked very well with the older version of AWX (19.5.1), and the issue started after we upgraded to AWX 21.9.0.

The issue was fixed in our new custom EE image, I'm not very sure if it's the upgrade of OS or `receptor` dependency or both did the trick.

For anyone who experienced the similar issue, I'd recommend checking the versions of your AWX and the EE image and possibly an upgrade of EE base image could help. 

Good luck!
